replay_buffer.py

```diff
@@ -327,35 +325,131 @@ def reanalyse(self, replay_buffer, shared_storage):
         ):
             self.model.set_weights(ray.get(shared_storage.get_info.remote("weights")))
+        
+        
+            # update target model periodically
+            if self.config.use_last_model_value:
+                training_step = ray.get(shared_storage.get_info.remote("training_step"))
+                if (training_step - self.last_update_step) >= self.config.value_target_update_freq :
+                    self.last_update_step = training_step
+                    self.target_model.set_weights(ray.get(shared_storage.get_info.remote("weights")))
 
-            game_id, game_history, _ = ray.get(
-                replay_buffer.sample_game.remote(force_uniform=True)
+            game_id, game_history = ray.get(
+                replay_buffer.reanalyse_sample_game.remote()
             )
 
             # Use the last model to provide a fresher, stable n-step value (See paper appendix Reanalyze)
             if self.config.use_last_model_value:
-                observations = [
-                    game_history.get_stacked_observations(
-                        i, self.config.stacked_observations
+                
+                
+                # use the lagging network (representation + value) to obtain updated targets
+                if not self.config.use_updated_mcts_value_targets:
+                    observations = [
+                        game_history.get_stacked_observations(
+                            i, self.config.stacked_observations
+                        )
+                        for i in range(len(game_history.root_values))
+                    ]
+
+                    observations = (
+                        torch.tensor(observations)
+                        .float()
+                        .to(next(self.model.parameters()).device)
+                    )
+                    values = models.support_to_scalar(
+                        # use lagging parameters
+                        self.target_model.initial_inference(observations)[0],
+                        self.config.support_size,
+                    )
+                    root_values = (
+                        torch.squeeze(values).detach().cpu().numpy()
                     )
-                    for i in range(len(game_history.root_values))
-                ]
 
-                observations = (
-                    torch.tensor(observations)
-                    .float()
-                    .to(next(self.model.parameters()).device)
-                )
-                values = models.support_to_scalar(
-                    self.model.initial_inference(observations)[0],
-                    self.config.support_size,
-                )
-                game_history.reanalysed_predicted_root_values = (
-                    torch.squeeze(values).detach().cpu().numpy()
-                )
+                # re-execute MCTS to update targets (child visist and root_values)
+                l = len(game_history.root_values)
```

Comment:
Hey,

Should there not be an else on this part?
If I understand correctly, the if in line 346 is there to trigger "do not use re-executed MCTS tree roots, use updated values directly from the value function instead".
So these lines should trigger only if the if is not firing, I believe
Apologies if the mistake is mine

Reply:
Ah I'm seeing now I missed the if in line 392, which solves the problem of updating with tree roots without intending to.
For computation efficiency though, wouldn't it still be better to add the else, instead of computing both?

Reply:
I believe there shouldn't be an "else" on this since we have to execute MCTS to obtain a "fresh" policy (used as target for training, 80% of the time as mentioned in the paper, MuZero Reanalyze appendix). Meanwhile the value function is either updated via re-running MCTS or from a target network.